### ISSUE TITLE
zend_mm_heap corrupted on updates or upserts with parallel extension

### ISSUE THREAD

**Issue by Totktonada** [tarantool/tarantool-php#152]
See https://github.com/tarantool/tarantool-php/pull/148#issuecomment-544684011 for the reproducer.

**Comment by Totktonada** [comment]
NB: Try to implement simple reproducer using `parallel` extension. Tried with `pthread` extension, but no luck.

**Comment by rybakit** [comment]
Yeah,  it's actually difficult to reproduce, and as I said in another issue, it may well be a bug in the `ext-parallel`:
1) https://github.com/krakjoe/parallel/issues/65
2) https://github.com/krakjoe/parallel/issues/129

**Comment by Totktonada** [comment]
It is a bit suspectful that the problems appears only with update and upsert.

**Comment by eljashm** [comment]
Reproducible even with two parallel workers. 
Tried to run under valgrind. There were some errors at the start but not relevant to the issue. Also, tried to run with USE_ZEND_ALLOC=0 which resulted in process exit with no exception.
